src/StreamField.js

```diff
@@ -0,0 +1,147 @@
+  children: PropTypes.arrayOf(lazyFunction(() => BlockDefinitionType)),
+});
+
+
+const BlockValueType = PropTypes.shape({
+  type: PropTypes.string.isRequired,
+  html: PropTypes.string,
+  hasError: PropTypes.bool,
+  value: PropTypes.oneOfType([
+    PropTypes.arrayOf(lazyFunction(() => BlockValueType)),
+    PropTypes.string,
+    PropTypes.number,
+    PropTypes.bool,
+  ]),
+});
+
+
+@connect((state, props) => {
+  const {id} = props;
+  const fieldData = state[id];
+  return {
+    generatedValue: fieldData === undefined ? '' : stateToValue(state, id),
+    isMobile: fieldData === undefined ? null : fieldData.isMobile,
+  };
+}, (dispatch, props) => {
+  const {id} = props;
+  return bindActionCreators({
+    initializeStreamField: data => initializeStreamField(id, data),
+    setIsMobile: isMobile => setIsMobile(id, isMobile),
+    moveBlock: (blockId, newIndex) => moveBlock(id, blockId, newIndex),
+  }, dispatch);
+})
+class StreamField extends React.Component {
+  static propTypes = {
+    id: PropTypes.string.isRequired,
+    required: PropTypes.bool,
+    minNum: PropTypes.number,
+    maxNum: PropTypes.number,
+    icons: PropTypes.shape({
+      moveUp: PropTypes.string,
+      moveDown: PropTypes.string,
+      duplicate: PropTypes.string,
+      delete: PropTypes.string,
+      grip: PropTypes.string,
+    }),
+    blockDefinitions: PropTypes.arrayOf(BlockDefinitionType).isRequired,
+    value: PropTypes.arrayOf(BlockValueType).isRequired,
+  };
+
+  static defaultProps = {
+    required: false,
+    minNum: 0,
+    maxNum: Infinity,
+  };
+
+  constructor(props) {
+    super(props);
+    const {
+      initializeStreamField, required, minNum, maxNum, icons, blockDefinitions,
+      value,
+    } = this.props;
+    initializeStreamField({
+      required, minNum, maxNum, icons, blockDefinitions,
+      isMobile: getIsMobile(),
+      value,
+    });
+    window.addEventListener('resize', this.onWindowResize);
+  }
+
+  onWindowResize = () => {
+    const value = getIsMobile();
+    if (value !== this.props.isMobile) {
+      this.props.setIsMobile(value);
+    }
+  };
+
+  onDragEnd = result => {
+    const {draggableId, source, destination} = result;
+    if (!destination || (result.reason === 'CANCEL')
+        || (destination.droppableId !== source.droppableId)
+        || (destination.index === source.index)) {
+      return;
+    }
+    this.props.moveBlock(draggableId, destination.index);
+  };
+
+  componentWillMount() {
+    // Removes the input with the same name if it exists.
+    const input = document.querySelector(`[name="${this.props.id}"]`);
+    if (input !== null) {
+      input.parentNode.removeChild(input);
+    }
+  }
+
+  render() {
+    const {id, generatedValue} = this.props;
+    return (
+      <DragDropContext onDragEnd={this.onDragEnd}>
+        <BlocksContainer fieldId={id} />
+        <input type="hidden" name={id}
+               value={JSON.stringify(generatedValue)} />
```

Comment:
I find this strange that `StreamField` takes the responsibility for storing its data like this, and patches over Wagtail's storage in its `componentWillMount`. I think it would be better if the component exposed an `onChange` callback (or similar) in its props, so Wagtail can take charge of storing the data.